polytracker/src/taintdag/fntrace.cpp

```diff
@@ -16,35 +16,16 @@ namespace taintdag {
 
 namespace {
 
-using offset_t = FnTrace::offset_t;
-using fn_index_t = FnMapping::index_t;
+using fn_index_t = Functions::index_t;
 
 } // namespace
 
-FnTrace::FnTrace(char *begin, char *end)
-    : trace_begin(begin), trace_end(end), events_end(trace_begin) {}
-
-std::optional<offset_t> FnTrace::write_event(event_t event) {
-  auto dst{events_end};
-  auto end{events_end + sizeof(event_t)};
-  if (end > trace_end) {
-    return {};
-  }
-  std::memcpy(dst, &event, sizeof(event_t));
-  return dst - trace_begin;
-}
-
-size_t FnTrace::get_event_count() {
-  return (events_end - trace_begin) / sizeof(event_t);
-}
-
-void FnTrace::log_fn_event(event_t::kind_t kind, fn_index_t idx) {
-  std::unique_lock write_lock{memory_m};
-  auto maybe_offset{write_event({kind, idx})};
-  if (!maybe_offset) {
-    error_exit("Failed to log event with id: ", get_event_count());
+void Events::log_fn_event(Event::kind_t kind, Functions::index_t idx) {
+  // Write an `Event` via `construct`
+  auto maybe_ctx{construct(kind, idx)};
```

Comment:
The `maybe_ctx` isn't really used, could just `if (!construct(kind, idx)) {` if you want to.